src/crimson/os/seastore/lba/btree_lba_manager.cc

```diff
@@ -665,45 +651,27 @@ BtreeLBAManager::search_insert_position(
   alloc_policy_t policy)
 {
   LOG_PREFIX(BtreeLBAManager::search_insert_position);
-  auto lookup_attempts = stats.num_alloc_extents_iter_nexts;
-  using OptIter = std::optional<LBABtree::iterator>;
-  return seastar::do_with(
-    hint, OptIter(std::nullopt),
-    [this, c, &btree, hint, length, lookup_attempts, policy, FNAME]
-    (laddr_t &last_end, OptIter &insert_iter)
-  {
-    return LBABtree::iterate_repeat(
-      c,
-      btree.upper_bound_right(c, hint),
-      [this, c, hint, length, lookup_attempts, policy,
-       &last_end, &insert_iter, FNAME](auto &iter)
-    {
-      ++stats.num_alloc_extents_iter_nexts;
-      if (iter.is_end() ||
-	  iter.get_key() >= (last_end + length)) {
-	if (policy == alloc_policy_t::deterministic) {
-	  ceph_assert(hint == last_end);
-	}
-	DEBUGT("hint: {}~0x{:x}, allocated laddr: {}, insert position: {}, "
-	       "done with {} attempts",
-	       c.trans, hint, length, last_end, iter,
-	       stats.num_alloc_extents_iter_nexts - lookup_attempts);
-	insert_iter.emplace(iter);
-	return search_insert_position_iertr::make_ready_future<
-	  seastar::stop_iteration>(seastar::stop_iteration::yes);
-      }
-      ceph_assert(policy == alloc_policy_t::linear_search);
-      last_end = (iter.get_key() + iter.get_val().len).checked_to_laddr();
-      TRACET("hint: {}~0x{:x}, current iter: {}, repeat ...",
-	     c.trans, hint, length, iter);
-      return search_insert_position_iertr::make_ready_future<
-	seastar::stop_iteration>(seastar::stop_iteration::no);
-    }).si_then([&last_end, &insert_iter] {
-      ceph_assert(insert_iter);
-      return search_insert_position_iertr::make_ready_future<
-	insert_position_t>(last_end, *std::move(insert_iter));
-    });
-  });
+  auto lookup_attempts = stats.num_alloc_extents_iter_nexts++;
+  auto iter = co_await btree.upper_bound_right(c, hint);
+  auto last_end = hint;
+
+  while (!iter.is_end() && iter.get_key() < (last_end + length)) {
+    ceph_assert(policy == alloc_policy_t::linear_search);
+    last_end = (iter.get_key() + iter.get_val().len).checked_to_laddr();
+    TRACET("hint: {}~0x{:x}, current iter: {}, repeat ...",
+	   c.trans, hint, length, iter);
+    iter = co_await iter.next(c);
+    ++stats.num_alloc_extents_iter_nexts;
+  }
+
+  if (policy == alloc_policy_t::deterministic) {
+    ceph_assert(hint == last_end);
+  }
+  DEBUGT("hint: {}~0x{:x}, allocated laddr: {}, insert position: {}, "
+	 "done with {} attempts",
+	 c.trans, hint, length, last_end, iter,
+	 stats.num_alloc_extents_iter_nexts - lookup_attempts);
+  co_return insert_position_t(last_end, std::move(iter));
```

Comment:
same here, how using while instead of LBABtree::iterate_repeat to handle interruption?